lib/vnet/setup_darwin.go

```diff
@@ -0,0 +1,254 @@
+	}
+	if err := os.Chmod(socketPath, 0600); err != nil {
+		return nil, "", trace.Wrap(err, "setting permissions on unix socket")
+	}
+	return l, socketPath, nil
+}
+
+// sendTUNNameAndFd sends the name of the TUN device and its open file descriptor over a unix socket, meant
+// for passing the TUN from the root process which must create it to the user process.
+func sendTUNNameAndFd(socketPath, tunName string, fd uintptr) error {
+	socketAddr := &net.UnixAddr{Name: socketPath, Net: "unix"}
+	conn, err := net.DialUnix(socketAddr.Net, nil /*laddr*/, socketAddr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer conn.Close()
+
+	err = conn.SetDeadline(time.Now().Add(tunHandoverTimeout))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Write the device name as the main message and pass the file desciptor as out-of-band data.
+	rights := unix.UnixRights(int(fd))
+	if _, _, err := conn.WriteMsgUnix([]byte(tunName), rights, socketAddr); err != nil {
+		return trace.Wrap(err, "writing to unix conn")
+	}
+	return nil
+}
+
+// sendTUNNameAndFd receives the name of a TUN device and its open file descriptor over a unix socket, meant
+// for passing the TUN from the root process which must create it to the user process.
+func recvTUNNameAndFd(ctx context.Context, socket *net.UnixListener) (string, uintptr, error) {
+	ctx, cancel := context.WithTimeout(ctx, tunHandoverTimeout)
+	defer cancel()
+	deadline, _ := ctx.Deadline()
+
+	err := socket.SetDeadline(deadline)
+	if err != nil {
+		return "", 0, trace.Wrap(err)
+	}
+	go func() {
+		<-ctx.Done()
+		socket.Close()
+	}()
+
+	conn, err := socket.AcceptUnix()
+	if err != nil {
+		return "", 0, trace.Wrap(err)
+	}
+	go func() {
+		// Close the connection early to unblock reads if the context is canceled.
+		<-ctx.Done()
+		conn.Close()
+	}()
+
+	msg := make([]byte, 128)
+	oob := make([]byte, unix.CmsgSpace(4)) // Fd is 4 bytes
+	n, oobn, _, _, err := conn.ReadMsgUnix(msg, oob)
+	if err != nil {
+		return "", 0, trace.Wrap(err, "reading from unix conn")
+	}
+
+	// Parse the device name from the main message.
+	if n == 0 {
+		return "", 0, trace.Errorf("failed to read msg from unix conn")
+	}
+	if oobn != len(oob) {
+		return "", 0, trace.Errorf("failed to read out-of-band data from unix conn")
+	}
+	tunName := string(msg[:n])
+
+	// Parse the file descriptor from the out-of-band data.
+	scm, err := unix.ParseSocketControlMessage(oob)
+	if err != nil {
+		return "", 0, trace.Wrap(err, "parsing socket control message")
+	}
+	if len(scm) != 1 {
+		return "", 0, trace.BadParameter("expect 1 socket control message, got %d", len(scm))
+	}
+	fds, err := unix.ParseUnixRights(&scm[0])
+	if err != nil {
+		return "", 0, trace.Wrap(err, "parsing file descriptors")
+	}
+	if len(fds) != 1 {
+		return "", 0, trace.BadParameter("expected 1 file descriptor, got %d", len(fds))
+	}
+	fd := uintptr(fds[0])
+
+	return tunName, fd, nil
+}
+
+func configureOS(ctx context.Context, cfg *osConfig) error {
```

Comment:
it could but I don't like that idea for some reason - osConfig is plain old configuration data, it doesn't need methods with huge side effects like modifying the operating system

Reply:
That reasoning doesn't make sense. The configuration data is exclusively used by this function in order to create huge side effects by modifying the operating system. It's just a matter of whether you use the language feature or not to pass it in to this method.

That said, its merely a nit, so I won't belabor it beyond pointing out that you haven't made a point.

Reply:
sure, I wasn't really trying to articulate a real point, just a feeling that this single plain old function taking a plain old struct is fine how it is. I would change it if you gave me a reason to, I care very little tbh

are you working on more comments or did you forget to approve by chance?